relayer/src/link/operational_data.rs

```diff
@@ -178,6 +200,155 @@ impl OperationalData {
 
         Ok(tm)
     }
+
+    /// Returns true iff the batch contains a packet event
+    fn has_packet_msgs(&self) -> bool {
+        self.batch.iter().any(|msg| msg.event.packet().is_some())
+    }
+
+    /// Returns the `connection_delay` iff the connection delay for this relaying path is non-zero
+    /// and the `batch` contains packet messages.
+    fn get_delay_if_needed(&self) -> Option<&ConnectionDelay> {
+        self.connection_delay
+            .as_ref()
+            .filter(|_| self.has_packet_msgs())
+    }
+
+    /// Returns `true` iff the connection delay for this relaying path is non-zero and `op_data`
+    /// contains packet messages.
+    pub fn conn_delay_needed(&self) -> bool {
+        self.get_delay_if_needed().is_some()
+    }
+
+    /// Sets the scheduled time that is used for connection-delay calculations
+    pub fn set_scheduled_time(&mut self, scheduled_time: Instant) {
+        if let Some(mut delay) = self.connection_delay.as_mut() {
+            delay.scheduled_time = scheduled_time;
+        }
+    }
+
+    /// Sets the update height that is used for connection-delay calculations
+    pub fn set_update_height(&mut self, update_height: Height) {
+        if let Some(mut delay) = self.connection_delay.as_mut() {
+            delay.update_height = Some(update_height);
+        }
+    }
+
+    /// Returns `Ok(remaining-delay)` on success or `LinkError` if the input closure fails.
+    pub fn conn_time_delay_remaining<ChainTime>(
```

Comment:
This and `conn_block_delay_remaining` can be made private.

I think of the privacy keywords more as a way to signal where a function is being used. If we see a private function, we know we only need to look through one file to find out where that function is used. While for public functions we have to search through the entire code base to figure out each of the use case.